src/shims/windows/thread.rs

```diff
@@ -0,0 +1,79 @@
+use std::time::{Duration, Instant};
+
+use rustc_middle::ty::layout::LayoutOf;
+use rustc_target::spec::abi::Abi;
+
+use crate::thread::Time;
+use crate::*;
+use shims::windows::handle::Handle;
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+
+#[allow(non_snake_case)]
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn CreateThread(
+        &mut self,
+        security_op: &OpTy<'tcx, Tag>,
+        stacksize_op: &OpTy<'tcx, Tag>,
+        start_op: &OpTy<'tcx, Tag>,
+        arg_op: &OpTy<'tcx, Tag>,
+        flags_op: &OpTy<'tcx, Tag>,
+        thread_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, ThreadId> {
+        let this = self.eval_context_mut();
+
+        if !this.ptr_is_null(this.read_pointer(security_op)?)? {
+            throw_unsup_format!("non-null `lpThreadAttributes` in `CreateThread`")
+        }
+
+        // stacksize is ignored, but still needs to be a valid usize
+        let _ = this.read_scalar(stacksize_op)?.to_machine_usize(this)?;
+
+        if this.read_scalar(flags_op)?.to_u32()? != 0 {
+            throw_unsup_format!("non-zero `dwCreationFlags` in `CreateThread`")
+        }
+
+        let thread =
+            if this.ptr_is_null(this.read_pointer(thread_op)?)? { None } else { Some(thread_op) };
+
+        this.start_thread(
+            thread,
+            start_op,
+            Abi::System { unwind: false },
+            arg_op,
+            this.layout_of(this.tcx.types.u32)?,
+        )
+    }
+
+    fn WaitForSingleObject(
+        &mut self,
+        handle: &OpTy<'tcx, Tag>,
+        timeout: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let thread = match Handle::from_scalar(this.read_scalar(handle)?.check_init()?, this)? {
+            Some(Handle::Thread(thread)) => thread,
+            Some(Handle::CurrentThread) => throw_ub_format!("trying to wait on itself"),
+            _ => throw_ub_format!("invalid handle"),
+        };
+
+        if this.read_scalar(timeout)?.to_u32()? != this.eval_windows("c", "INFINITE")?.to_u32()? {
+            this.check_no_isolation("`WaitForSingleObject` with non-infinite timeout")?;
+        }
+
+        let timeout_ms = this.read_scalar(timeout)?.to_u32()?;
+
+        let timeout_time = if timeout_ms == this.eval_windows("c", "INFINITE")?.to_u32()? {
+            None
+        } else {
+            let duration = Duration::from_millis(timeout_ms as u64);
```

Comment:
Please make sure that timeouts only work when isolation is disabled. Thus is important: Miri in its default flags should be fully deterministic, and real-time clocks are not deterministic.

Also, is there a test covering this? Can the Rust stdlib even join a thread with a timeout?